src/tracks/Pattern.cpp

```diff
@@ -295,10 +290,21 @@ void Pattern::clearNotes()
 
 
 
-Note * Pattern::addStepNote( int step )
+Note * Pattern::addStepNote(int step)
 {
-	return addNote( Note( TimePos( -DefaultTicksPerBar ),
-				TimePos::stepPosition( step ) ), false );
+	Note * n =
+		addNote(
+			Note(TimePos(DefaultTicksPerBar / 16),
+			TimePos::stepPosition(step)), false
+		);
+	n->setType(Note::StepNote);
+	// Have to check the type again now that we changed
+	// the type of the note
+	// TODO: This means checkType is called twice:
+	// Once on addNote and once here. Rethink it so we
+	// can call it only once.
+	checkType();
```

Comment:
If `m_patternType` was `MelodyPattern` before `addNote`, then it will stay `MelodyPattern` afterwards.
If `m_patternType` was `BeatPattern` before `addNote`, then it will stay `BeatPattern` afterwards.

Therefore, if we grab `m_patternType` before calling `addNote`, we should be able to simply know if we need to keep the same type or change.

```suggestion
	auto prev_patternType = m_patternType;
	Note * n =
		addNote(
			Note(TimePos(DefaultTicksPerBar / 16),
			TimePos::stepPosition(step)), false
		);
	n->setType(Note::StepNote);
	// addNote will change to MelodyPattern with addNote (calls checkType),
	// we need to keep it BeatPattern if it was beforehand
	if (prev_patternType == BeatPattern) { setType(BeatPattern); }
```

Reply:
That's a good idea! I'll change the code to use it